internal/impl/sql/dynamic_credential_wrapper.go

```diff
@@ -0,0 +1,106 @@
+package sql
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"errors"
+	"github.com/benthosdev/benthos/v4/public/service"
+	"net/url"
+	"strings"
+	"sync"
+)
+
+type DynamicCredentialConnector struct {
+	mu                 sync.Mutex
+	driver             driver.Driver
+	driverName         string
+	dsn                *service.InterpolatedString
+	dynamicCredentials *DynamicCredentials
+	manager            *service.Resources
+}
+
+func NewDynamicCredentialConnector(manager *service.Resources, driverName string, dsn *service.InterpolatedString, dynamicCredentials *DynamicCredentials) (*DynamicCredentialConnector, error) {
+	driveri, err := sql.Open(driverName, "")
+	if err != nil {
+		return nil, err
+	}
+	resolvedDriver := driveri.Driver()
+	_ = driveri.Close()
```

Comment:
This is a bit of a hack to get the underlying driver using our driver string. The stdlib sql package specified that no connections should be created when calling open, so this should be generally safe. Tests for all current drivers are written and pass.